### ISSUE TITLE
Redesign Notifications

### ISSUE THREAD

**Issue by pushcx** [lobsters/lobsters#1076]
Notifications on Lobsters have slowly grown into a fairly hairy muddle. There is inconsistent UI with odd limitations based on vintage. There's a good-sized project here if someone would like to take it on. In particular, this could make a nice portfolio piece for a junior dev.

Things people get notified about

 * Top-level comments to stories they've submitted
 * Direct to their comments (not grandchild replies)
 * `@username` mentions
 * Private messages

Perhaps also users who are particularly interested in a story or response would like to be able to subscribe to it the same way the author can.

Channels people get notified via:

 * Onsite [`/replies`](https://lobste.rs/replies)
 * Onsite [`/messages`](https://lobste.rs/messages)
 * E-mail
 * Pushover
 * [Your Threads](https://lobste.rs/threads)

There's no real rhyme or reason to which things generate which notifications. Except for the distinction between comments and PMs that shows up in `/replies` and `/comments`, maybe all channels should be available for all things?

And speaking of that exception, maybe we should unify the interface for new PMs and new
replies. I'm a fan of the way Reddit presents them together and nudged this direction a
little with the new header's singular Inbox link (#1061).

Additionally, `/replies` works as a view via the `ReplyingComments` model. While an excellent feature with [nice limitations](https://github.com/lobsters/lobsters/blob/e49e1672d7675dd4e1f193bcf3cb1f17939464d4/db/views/replying_comments_v09.sql#L31-L52) to turn down flame wars, it's the most expensive query the site regularly runs (and addressing that with caching [prompted several bugs](https://github.com/lobsters/lobsters/issues/1065)). It would be best if the view's performance could be improved (see `EXPLAIN` below), but maybe this needs to be entirely rebuilt with more of a materialized design? (A hassle there is that some of the conditions that exclude a comment like flags can come well after the comment was posted.)

```
EXPLAIN SELECT     read_ribbons.user_id,     comments.id as comment_id,     read_ribbons.story_id as story_id,     comments.parent_comment_id,     comments.created_at as comment_created_at,     parent_comments.user_id as parent_comment_author_id,     comments.user_id as comment_author_id,     stories.user_id as story_author_id,     (read_ribbons.updated_at < comments.created_at) as is_unread,     (select votes.vote from votes where votes.user_id = read_ribbons.user_id and votes.comment_id = comments.id) as current_vote_vote,     (select votes.reason from votes where votes.user_id = read_ribbons.user_id and votes.comment_id = comments.id) as current_vote_reason FROM     read_ribbons JOIN     comments ON comments.story_id = read_ribbons.story_id JOIN     stories ON stories.id = comments.story_id LEFT JOIN     comments parent_comments ON parent_comments.id = comments.parent_comment_id WHERE     read_ribbons.is_following = 1     AND comments.user_id != read_ribbons.user_id     AND comments.is_deleted = FALSE     AND comments.is_moderated = FALSE     AND (       parent_comments.user_id = read_ribbons.user_id       OR (parent_comments.user_id IS NULL       AND stories.user_id = read_ribbons.user_id)     )          AND stories.score >= 0     AND comments.score >= 0     AND (       parent_comments.id IS NULL       OR (                  parent_comments.score >= 0         AND parent_comments.is_moderated = FALSE         AND parent_comments.is_deleted = FALSE       )     )     AND (        NOT EXISTS (         SELECT 1 FROM votes f JOIN comments c ON f.comment_id = c.id         WHERE           f.vote < 0           AND f.user_id = parent_comments.user_id           AND c.user_id = comments.user_id           AND f.story_id = comments.story_id       )     );
+------+--------------------+-----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+----------------------------------------------------+--------+-------------+
| id   | select_type        | table           | type   | possible_keys                                                                                                                                         | key                            | key_len | ref                                                | rows   | Extra       |
+------+--------------------+-----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+----------------------------------------------------+--------+-------------+
|    1 | PRIMARY            | stories         | ALL    | PRIMARY,index_stories_on_user_id,index_stories_on_score,index_stories_on_id_and_is_deleted                                                            | NULL                           | NULL    | NULL                                               | 84845  | Using where |
|    1 | PRIMARY            | comments        | ref    | story_id_short_id,index_comments_on_score                                                                                                             | story_id_short_id              | 8       | lobsters.stories.id                                | 4      | Using where |
|    1 | PRIMARY            | parent_comments | eq_ref | PRIMARY                                                                                                                                               | PRIMARY                        | 8       | lobsters.comments.parent_comment_id                | 1      | Using where |
|    1 | PRIMARY            | read_ribbons    | ref    | index_read_ribbons_on_user_id,index_read_ribbons_on_story_id                                                                                          | index_read_ribbons_on_story_id | 8       | lobsters.stories.id                                | 50     | Using where |
|    4 | MATERIALIZED       | c               | index  | PRIMARY,thread_id,story_id_short_id,confidence_idx,index_comments_on_user_id,comments_hat_id_fk,comments_parent_comment_id_fk,index_comments_on_score | index_comments_on_user_id      | 8       | NULL                                               | 393828 | Using index |
|    4 | MATERIALIZED       | f               | ref    | user_id_comment_id,user_id_story_id,index_votes_on_comment_id,votes_story_id_fk                                                                       | index_votes_on_comment_id      | 9       | lobsters.c.id                                      | 3      | Using where |
|    3 | DEPENDENT SUBQUERY | votes           | ref    | user_id_comment_id,user_id_story_id,index_votes_on_comment_id                                                                                         | user_id_comment_id             | 17      | lobsters.read_ribbons.user_id,lobsters.comments.id | 1      |             |
|    2 | DEPENDENT SUBQUERY | votes           | ref    | user_id_comment_id,user_id_story_id,index_votes_on_comment_id                                                                                         | user_id_comment_id             | 17      | lobsters.read_ribbons.user_id,lobsters.comments.id | 1      |             |
+------+--------------------+-----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+----------------------------------------------------+--------+-------------+
```

**Comment by thomasdziedzic** [comment]
I'm going to try to tackle this.as part of #539 

Since this was last updated over 3 years ago, are there any changes to the original description in this issue?

I'm going to describe a high level system that is probably going to change but wanted to get your feedback on this to see where (if any) high level misunderstandings are.

At first glance I was imagining the following components:
1. subscription: used to associate a user to either a comment or story. This could be used to implement an unsubscribe feature in the future. This could also be used to implement user subscriptions by non-authors in the future.
2. subscription_entry: used to track comments to subscribed objects, and tracks read/unread notifications. Also used to represent @ mentions and private message events (but not the actual private messages themselves).
3. user_channel: represents the different configured channels for a user which can be scoped to subscription types

The way this would work together is an event (comment, @ mention, PM) finds subscriptions if relevant and creates a subscription_entries for each subscription. There would be a background job that looks for subscription_entries and sends notifications per user_channel configured for a user.

**Comment by pushcx** [comment]
Great to have an experienced community member and developer take on a pretty vaguely-written up feature. Your plan looks like it'll hit all the features we want to allow general preferences and then toggle on or off subscriptions for particular items.

@caius suggested the [noticed](https://github.com/excid3/noticed/) gem that looks like it does a lot of what we want here if we put the subscription/channel model that @thomasdziedzic sketched out into the `before_enqueue` hook. It looks like a well-maintained and documented dep that I'd be willing to take because it'll roll up a whole lot of feature code that's shotgunned through controllers and (shakes fist at cloud) model callbacks.

**Comment by thomasdziedzic** [comment]
Just leaving a reference here, #1624 was done as part of this work to move the notifications from an after create hook, out into an ActiveJob.

**Comment by thomasdziedzic** [comment]
Tried to write a script to generate users:

```
generated 18538 users in 27479 seconds at a rate of 0 records per second
```

This generated users at a rate of 0.67 records per second which took 7.5 hours to finish.
I'll see if comments are just as bad, but if so I'll probably look into using the parallel gem since it [mentions activerecord](https://github.com/grosser/parallel?tab=readme-ov-file#activerecord) as one of the uses.

**Comment by pushcx** [comment]
If you still have the script, it might be worth a second run wrapped with `User.transaction do ... end`, and call `save(validate: false)` to skip validations. I'd expect it to run a lot faster but maybe it's worth checking that with 1k users before you maybe spend a day on it.

**Comment by thomasdziedzic** [comment]
I tried using the parallel gem and kept getting a deadlock for some reason, I didn't debug it further.

My approach is to just write a script and let it run for x hours since I don't have to be around the computer, I've already got comments generated using create! which took 21 hours. I still have about 3M votes to generate and 400k read_ribbons which should take care of all the "big" tables. After each "step" I take a database backup in case I mess something up or I want to redo something.

Wrapping with a transaction and skipping validations does sound like it would save some time, I'll give that a shot for read_ribbons and compare performance.

**Comment by thomasdziedzic** [comment]
I finished waiting for the fake bulk data script and generated about 2.5G of data over the past few days. Not sure if there's a good place of uploading this amount of data publicly for others to use.

I created a branch to keep this script under: https://github.com/lobsters/lobsters/compare/master...thomasdziedzic:lobsters:fake-bulk-data-script

**Comment by thomasdziedzic** [comment]
I saw a message in the replies: "Skipped slow queries for flag warning in mode development"

<img width="704" alt="Image" src="https://github.com/user-attachments/assets/15e8519c-53e6-4524-8b89-8bea4ac4e81f" />

I found the logic where it displays that messages, but I couldn't find the logic where it skips the queries in the replies_controller. Do you know how development skips those slow queries?

**Comment by thomasdziedzic** [comment]
> I saw a message in the replies: "Skipped slow queries for flag warning in mode development"
> 
> <img alt="Image" width="704" src="https://private-user-images.githubusercontent.com/203340/458516488-15e8519c-53e6-4524-8b89-8bea4ac4e81f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA4MDYxMzUsIm5iZiI6MTc1MDgwNTgzNSwicGF0aCI6Ii8yMDMzNDAvNDU4NTE2NDg4LTE1ZTg1MTljLTUzZTYtNDUyNC04Yjg5LThiZWE0YWM0ZTgxZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyNFQyMjU3MTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZDBkYmIxY2I5M2UwN2Q3Mzg3NWY0ZmRlNmI0ZTNhZTRmMjU2MzY5NTFkNWY3YjExNTgxNmIxN2VlY2E3MTA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Nq5xqmcGD6cM-U9GJqKIMW4TrPKu1FYA5dh6VJRmOL4">
> I found the logic where it displays that messages, but I couldn't find the logic where it skips the queries in the replies_controller. Do you know how development skips those slow queries?

I think that warning about skipped slow queries might be outdated, I didn't find any skip logic and root caused the lack of replies to me generating random read_ribbons rather than creating read_ribbons for all of the story and comment authors. Once I added that, the queries still ran, displayed the replies, but also still displayed the info. Maybe that info was once correct but it no longer applies? Let me know what you think.

**Comment by pushcx** [comment]
That warning is about the `possible_flag_warning` helper in `app/helpers/application_helper.rb` (talked a bit more about this on [today's office hours](https://push.cx/stream/2025-06-26).

**Comment by thomasdziedzic** [comment]
From the original description:

> And speaking of that exception, maybe we should unify the interface for new PMs and new
> replies. I'm a fan of the way Reddit presents them together and nudged this direction a
> little with the new header's singular Inbox link (https://github.com/lobsters/lobsters/pull/1061).

Can you explain what you mean by a unified interface for PMs and new replies? Do you mean private messages and comments should be displayed in the same list ideally? Sorry, I don't really use reddit, only to get links sometimes.

In addition:

> It would be best if the view's performance could be improved (see EXPLAIN below), but maybe this needs to be entirely rebuilt with more of a materialized design?

Do you remember where you encountered the performance issues? I clicked on the inbox tabs for various users that use the view, and I'm getting double digit ms response times which seems pretty good.

**Comment by pushcx** [comment]
Yes, that's what reddit looks like. There's not a lot of UI difference between the two; we'd have to do a `UNION` query and Rails doesn't have great support for those, on top of how the performance issues with the ReplyingComments view. But if there's a single Notifications table, both issues evaporate.

If you didn't see perf issues, your data isn't big enough or in some other way doesn't match prod. It's slow, typically taking 1100ms to run in prod, that's why I cached the `inbox_count` and `unread_replies_count`. Slow enough that if we'd had SolidQueue working I would've shoved that into a background job and served a state number/0 until it filled.

**Comment by thomasdziedzic** [comment]
I have an initial pass at adding a UI for notifications:

<img width="701" alt="Image" src="https://github.com/user-attachments/assets/44399c9c-11b1-48bf-86c1-34eda4066154" />

Let me know what you think!

The top notification has a bolded subject to indicate that it's unread.

**Comment by pushcx** [comment]
This is exactly the kind of combined view I was hoping for. This looks really good.

The bold was too subtle for me to recognize, it took fullscreening and zooming to get it. Maybe take the red `<span class='comment_unread'>` from `app/views/comments/_comment.html.erb`? (I guess it could just be named 'unread'.) The js for clicking '(unread)' and jumping through the page would even work here, which is a satisfying bit of feature consistency.

For the Messages, what if the envelope icon was in the left gutter in the place of the upvoter? That'd let it be a bit bigger, too. The folder control on comments would probably be a confusingly similar rectangle, but it doesn't do anything useful on this page so could be removed. I guess I'm saying I really like these together, but don't want to set people up to make an embarrassing mistake about public vs private. If the Comment 'reply' button opens the inline reply and the Message 'reply' button redirects over to that form, the flow is different enough that people shouldn't make it, but it's even better if they can tell at a glance.

I'm a big fan of incremental progress, does this UI feel like a coherent chunk you could PR?

**Comment by thomasdziedzic** [comment]
> This is exactly the kind of combined view I was hoping for. This looks really good.

Glad to hear that!

> The bold was too subtle for me to recognize, it took fullscreening and zooming to get it. Maybe take the red `<span class='comment_unread'>` from `app/views/comments/_comment.html.erb`? (I guess it could just be named 'unread'.) The js for clicking '(unread)' and jumping through the page would even work here, which is a satisfying bit of feature consistency.

That sounds like a good idea, I'll do that.

> For the Messages, what if the envelope icon was in the left gutter in the place of the upvoter? That'd let it be a bit bigger, too.

I had the same thought about placing the envelope in the left gutter. I'll do that as well.

> The folder control on comments would probably be a confusingly similar rectangle, but it doesn't do anything useful on this page so could be removed.

I'll remove it

> I guess I'm saying I really like these together, but don't want to set people up to make an embarrassing mistake about public vs private. If the Comment 'reply' button opens the inline reply and the Message 'reply' button redirects over to that form, the flow is different enough that people shouldn't make it, but it's even better if they can tell at a glance.

Yes that's how it is, I just noticed that the inline comment reply is broken on the notifications page so I'll fix that, but it should work as you described it.

> I'm a big fan of incremental progress, does this UI feel like a coherent chunk you could PR?

There was a lot of detail in this issue and I was going to PR it up altogether but I suppose a smaller PR makes this more manageable.

Currently the branch is still very much a work in progress but I can share it here: https://github.com/lobsters/lobsters/compare/master...thomasdziedzic:lobsters:notifications-plus-plus

It includes:
1. Subscriptions which decouple the author of a story/comment from subscribing to notifications.
2. Notifications which abstract the things you can get notified for which are what you see in the above screenshot.
3. A script to generate bulk data (takes a couple of days to finish unfortunately).
4. A script to backfill subscriptions with a story/comment author and notifications for those subscriptions. 

The branch does not include:
1. Any work related to channels (currently email / pushover) and just reuses existing logic.
2. Filter logic matching ReplyingComment for the backfill script or ongoing notifications.

Note that I technically can split out subscriptions work from notifications work since notifications will just use the existing logic to get created. That's always an option if you want.

**Comment by thomasdziedzic** [comment]
> > The bold was too subtle for me to recognize, it took fullscreening and zooming to get it. Maybe take the red `<span class='comment_unread'>` from `app/views/comments/_comment.html.erb`? (I guess it could just be named 'unread'.) The js for clicking '(unread)' and jumping through the page would even work here, which is a satisfying bit of feature consistency.
> 
> That sounds like a good idea, I'll do that.

This was an easy add, and like you stated, it "just worked" once it was added!

> > For the Messages, what if the envelope icon was in the left gutter in the place of the upvoter? That'd let it be a bit bigger, too.
> 
> I had the same thought about placing the envelope in the left gutter. I'll do that as well.

Added the envelope to the gutter and it looks much more discernable.

> > The folder control on comments would probably be a confusingly similar rectangle, but it doesn't do anything useful on this page so could be removed.
> 
> I'll remove it

Removed.

> > I guess I'm saying I really like these together, but don't want to set people up to make an embarrassing mistake about public vs private. If the Comment 'reply' button opens the inline reply and the Message 'reply' button redirects over to that form, the flow is different enough that people shouldn't make it, but it's even better if they can tell at a glance.
> 
> Yes that's how it is, I just noticed that the inline comment reply is broken on the notifications page so I'll fix that, but it should work as you described it.

I fixed the inline commenting but also removed the reply link for messages and added a link in the "time_ago" to be linked to the actual message which includes a reply to component. I think this is a more consistent way of doing it.

> > I'm a big fan of incremental progress, does this UI feel like a coherent chunk you could PR?

Opened a PR to get early feedback around this feature: https://github.com/lobsters/lobsters/pull/1635

**Comment by pushcx** [comment]
I deployed #1635 a few minutes ago. Backfilling the comment notifications took 843 seconds and the message notifications 31s.

I made some minor css/ruby style tweaks in the merge commit 827f9115. Jujutsu shows my changes when given the change id (okzxwutxswuwwzrqznvyzsxmukuqxxsw) but GitHub presents them and all the changes from the branch together, which is unfortunate. If I tweak another PR at merge I'll either put my changes in a standalone commit to make a 3-way merge or merge the PR cleanly and put my changes in a commit on top.

I reverted the change I asked for about marking things unread, sorry for the churn. When I played with the branch locally I realized being able to look at the notifications without clearing the 'unread' flag was pretty nice.

I also removed notifications from the message subnav. So users can visit https://lobste.rs/notifications to see it with real data but we can edit before fully rolling it out. I posted the link in [chat](https://lobste.rs/chat) to ask for feedback.

Here's my todo list from the PR for removing `/replies`.

 - [x] remove the `ReplyingComments` model and spec
 - [x] remove `User#unread_replies_count`, `clear_unread_replies!`, `inbox_count`, `unread_message_count`, and `update_unread_message_count`
 - [x] a migration to remove the `replying_comments` view and all `keystores` entries matching `user:.*:unread_messages` or `user:.*:unread_replies`
 - [x] rm `db/views`
 - [x] remove the `scenic` gem and bundle (I'd rather the previous migration execute raw sql `drop view` than split this up into two PRs)
 - [x] update the `inbox` link - actually, would this feature be nicer under `/inbox` than `/notifications`?
 - [x] add anti-flamewar predicates, see below
 - [x] redirect routes under `/replies`

That'll be enough to unblock a migration to sqlite (#539), it doesn't need to wait on the other features requested up top here.

**Comment by thomasdziedzic** [comment]
Could you elaborate what you mean by: "update the inbox link - actually, would this feature be nicer under /inbox than /notifications?" and "redirect routes under /replies"

Happy to do it, I just need to understand:
1. do you mean clicking the inbox should take you to unread notifications if there are any or all notifications if there aren't any unread?
2. redirect /replies to where? /notifications?

Opened #1649 since I had that as a stacked commit in JJ :)

**Comment by pushcx** [comment]
jjhub when, td? WHEN!?


By 'inbox link' I'm referring to [this link](https://github.com/lobsters/lobsters/blob/22eb74ae742790d64674c7ad40c29a5d3744c090/app/views/layouts/application.html.erb#L70) up in the sitewide header. Right now it links to unread messages if there are any, or unread replies otherwise. But now we have a unified view!

I have managed to typo `lobste.rs/notifications` 3 times since deploy, so I'm liking the idea of renaming to `/inbox` even more. :p

And, yeah, when ReplyingComments is gone, /replies URLs would be broken, it'll have to redirect to NotificationsController.

I'm concerned by the placeholder for hidden notifications. Even after I chose a less-bright emoji, they still seem intrusive. The idea of the filters in the replying_comments view was to avoid pouring gas on a flamewar by redirecting people to it, but every time I see one I go, "Oh ho, what drama happened here?" and want to go find the context. I'm mulling ways to improve them.

**Comment by thomasdziedzic** [comment]
> jjhub when, td? WHEN!?

haha, it's really not a bad idea. I wonder if the jj folks would mind.

> By 'inbox link' I'm referring to [this link](https://github.com/lobsters/lobsters/blob/22eb74ae742790d64674c7ad40c29a5d3744c090/app/views/layouts/application.html.erb#L70) up in the sitewide header. Right now it links to unread messages if there are any, or unread replies otherwise. But now we have a unified view!

Ah, in that case I think I have included that logic at https://github.com/lobsters/lobsters/pull/1649/files#diff-607cdd73a589ce835bd378a9a5d8480f96e0f82f04d9c79c89c7184f41d7283d which matches the same logic as was previously, just for notifications.

> I have managed to typo `lobste.rs/notifications` 3 times since deploy, so I'm liking the idea of renaming to `/inbox` even more. :p

I typically use links so I don't have this issue. But I can see how changing old habits of typing in a new url can affect some people. I moved the notifications controller/routes into the inbox_controller.

> And, yeah, when ReplyingComments is gone, /replies URLs would be broken, it'll have to redirect to NotificationsController.

Redirected the /replies to the new /inbox/all & /inbox/unread paths.

> I'm concerned by the placeholder for hidden notifications. Even after I chose a less-bright emoji, they still seem intrusive. The idea of the filters in the replying_comments view was to avoid pouring gas on a flamewar by redirecting people to it, but every time I see one I go, "Oh ho, what drama happened here?" and want to go find the context. I'm mulling ways to improve them.

"want to go find the context" you mean this placeholder boosts user engagement? :)
/me takes off marketing hat

I can see how the placeholder could have the opposite effect, though here is the way I see it, we can either:
1. Have placeholders to show that we have 25 notifications per page. The more I think about it, the less appealing this seems in the long run since as you said, it seems to have the opposite effect.
2. Accept that there might be less than 25 notifications in good faith displayed on the /inbox/all path
3. Have a database attribute that we maintain, something like `is_good_faith` where we put in hooks or a job to recalculate this new database attribute in all the places that can modify how we derive the current filter criteria. I was hoping to avoid this since this seemed like a lot of work to add and get right. We would have to recalculate, when a story, comment, vote, hidden_story, or filtered tag gets created, updated, or deleted. On the plus side, the pagination implementation would then be trivial given a database attribute.
4. Another option could be to just keep querying the database in a loop until we get 25 notifications or we reach the end, and build up 25 notifications that way. I'm not a fan of this one though because of the db query in a loop + pagination would be the same operation, just dropping `page * page_size` notifications first before being able to start `taking`.

**Comment by pushcx** [comment]
If you're seriously considering making a jjhub, the thing that needs the most work is PRs. The GH UI is clunky but probably frozen in amber because of popularity. I'm endlessly irritated by comments and in-diff comments being so separate, paging back and forth between tabs trying to reconcile what's happened over the life of the PR. It gets significantly harder if people revise commits. A while ago there I saw a [talk go by](https://lobste.rs/s/slhe5j/how_jane_street_does_code_review_2017) about a radically different workflow; I don't know that I'd prefer that particular workflow, but it's good to see someone take a big step away from the familiar. (And to store the discussion in the repo.)


There's more options:

 5. Option 3, but instead of doing it every time one of the contributing objects changes, do it when the user loads `/inbox`.
 6. Add a scope to notifications to filter them out. The `replying_comments` view did it in SQL, so it's certainly possible. Perf may return as an issue, but given that the query would start from the user's `notifications` (small) rather than `comments` (big and wide), maybe it'd be ok?
 7. Do what tests we can when creating the notification and accept that we may have chosen differently for some by the time we're read.

Zooming out to look at the other desired features, we already accept that email and pushover effectively take option 7 here. (Though really I overlooked them and they don't check at all. They should.)

So we're trading off these characteristics:

 * Query performance for this view
 * Write performance for commenting and voting
 * UI consistency with number per page
 * UI consistency with notifications that can appear and later disappear
 * Effectiveness of flamewar prevention

The last two characteristics are always going to be in conflict. Deferring the decision until the reader views their notifications is why `replying_comments` can be so expressive. It's been 7.5 years since we added the view and site norms have shifted away from flamewars in that time. Maybe it's safe to give up some effectiveness for everything else.

**Comment by thomasdziedzic** [comment]
> Maybe it's safe to give up some effectiveness for everything else.

I like this, would be good to know the effectiveness of each clause (what percentage of rows are they filtering out). Maybe there's some so small that we can just simplify the good_faith conditions by removing some of the less effective ones. I wrote up a query which might take a while but it at least starts running on my machine. It determines how effective each of the conditions is.

```sql
with base as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
),
story_score as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND stories.score > stories.flags
),
comment_score as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND comments.score > comments.flags
),
comment_is_not_deleted as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND comments.is_deleted = FALSE
),
comment_is_not_moderated as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND comments.is_moderated = FALSE
),
parent_score as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND (parent_comments.id IS NULL OR parent_comments.score > parent_comments.flags)
),
parent_is_not_deleted as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND (parent_comments.id IS NULL OR parent_comments.is_deleted = FALSE)
),
parent_is_not_moderated as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND (parent_comments.id IS NULL OR parent_comments.is_moderated = FALSE)
),
replier_not_flagged as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND (NOT EXISTS (SELECT 1 FROM votes f JOIN comments c ON f.comment_id = c.id WHERE f.vote < 0 AND f.user_id = parent_comments.user_id AND c.user_id = comments.user_id AND f.story_id = comments.story_id))
),
story_not_hidden as (
  SELECT count(*) cnt FROM read_ribbons JOIN comments ON comments.story_id = read_ribbons.story_id JOIN stories ON stories.id = comments.story_id LEFT JOIN comments parent_comments ON parent_comments.id = comments.parent_comment_id
  WHERE read_ribbons.is_following = 1 AND comments.user_id != read_ribbons.user_id AND (parent_comments.user_id = read_ribbons.user_id OR (parent_comments.user_id IS NULL AND stories.user_id = read_ribbons.user_id))
  AND (NOT EXISTS (SELECT 1 FROM hidden_stories h WHERE h.story_id = stories.id AND h.user_id = parent_comments.user_id))
)
select
  base.cnt - story_score.cnt                           as filtered_by_story_score,
  (base.cnt - story_score.cnt) / base.cnt              as filtered_by_story_score_pct,

  base.cnt - comment_score.cnt                         as filtered_by_comment_score,
  (base.cnt - comment_score.cnt) / base.cnt            as filtered_by_comment_score_pct,

  base.cnt - comment_is_not_deleted.cnt                as filtered_by_comment_is_not_deleted,
  (base.cnt - comment_is_not_deleted.cnt) / base.cnt   as filtered_by_comment_is_not_deleted_pct,

  base.cnt - comment_is_not_moderated.cnt              as filtered_by_comment_is_not_moderated,
  (base.cnt - comment_is_not_moderated.cnt) / base.cnt as filtered_by_comment_is_not_moderated_pct,

  base.cnt - parent_score.cnt                          as filtered_by_parent_score,
  (base.cnt - parent_score.cnt) / base.cnt             as filtered_by_parent_score_pct,

  base.cnt - parent_is_not_deleted.cnt                 as filtered_by_parent_is_not_deleted,
  (base.cnt - parent_is_not_deleted.cnt) / base.cnt    as filtered_by_parent_is_not_deleted_pct,

  base.cnt - parent_is_not_moderated.cnt               as filtered_by_parent_is_not_moderated,
  (base.cnt - parent_is_not_moderated.cnt) / base.cnt  as filtered_by_parent_is_not_moderated_pct,

  base.cnt - replier_not_flagged.cnt                   as filtered_by_replier_not_flagged,
  (base.cnt - replier_not_flagged.cnt) / base.cnt      as filtered_by_replier_not_flagged_pct,

  base.cnt - story_not_hidden.cnt                      as filtered_by_story_not_hidden,
  (base.cnt - story_not_hidden.cnt) / base.cnt         as filtered_by_story_not_hidden_pct
from base
cross join story_score
cross join comment_score
cross join comment_is_not_deleted
cross join comment_is_not_moderated
cross join parent_score
cross join parent_is_not_deleted
cross join parent_is_not_moderated
cross join replier_not_flagged
cross join story_not_hidden
;
```

**Comment by pushcx** [comment]
```
                 filtered_by_story_score: 1059
             filtered_by_story_score_pct: 0.0109
               filtered_by_comment_score: 1568
           filtered_by_comment_score_pct: 0.0162
      filtered_by_comment_is_not_deleted: 1586
  filtered_by_comment_is_not_deleted_pct: 0.0164
    filtered_by_comment_is_not_moderated: 336
filtered_by_comment_is_not_moderated_pct: 0.0035
                filtered_by_parent_score: 1233
            filtered_by_parent_score_pct: 0.0127
       filtered_by_parent_is_not_deleted: 479
   filtered_by_parent_is_not_deleted_pct: 0.0049
     filtered_by_parent_is_not_moderated: 251
 filtered_by_parent_is_not_moderated_pct: 0.0026
         filtered_by_replier_not_flagged: 628
     filtered_by_replier_not_flagged_pct: 0.0065
            filtered_by_story_not_hidden: 36
        filtered_by_story_not_hidden_pct: 0.0004
1 row in set (38 min 55.29 sec)
```

Flags are uncommon and I tried to make the conditions conservative, so it's not surprising these numbers are small. I'd love to compare the predicates at the time the comment was posted (or a minute later) against the time the notification was read. We can't exactly get it because of design choices like removing a vote/flag results in the record being deleted, but that's pretty rare so the reduction to accuracy should be minimal.

**Comment by thomasdziedzic** [comment]
> ```
>                  filtered_by_story_score: 1059
>              filtered_by_story_score_pct: 0.0109
>                filtered_by_comment_score: 1568
>            filtered_by_comment_score_pct: 0.0162
>       filtered_by_comment_is_not_deleted: 1586
>   filtered_by_comment_is_not_deleted_pct: 0.0164
>     filtered_by_comment_is_not_moderated: 336
> filtered_by_comment_is_not_moderated_pct: 0.0035
>                 filtered_by_parent_score: 1233
>             filtered_by_parent_score_pct: 0.0127
>        filtered_by_parent_is_not_deleted: 479
>    filtered_by_parent_is_not_deleted_pct: 0.0049
>      filtered_by_parent_is_not_moderated: 251
>  filtered_by_parent_is_not_moderated_pct: 0.0026
>          filtered_by_replier_not_flagged: 628
>      filtered_by_replier_not_flagged_pct: 0.0065
>             filtered_by_story_not_hidden: 36
>         filtered_by_story_not_hidden_pct: 0.0004
> 1 row in set (38 min 55.29 sec)
> ```

Thanks for running the query, it took a lot less time than I imagined it would.

> Flags are uncommon and I tried to make the conditions conservative, so it's not surprising these numbers are small. I'd love to compare the predicates at the time the comment was posted (or a minute later) against the time the notification was read. We can't exactly get it because of design choices like removing a vote/flag results in the record being deleted, but that's pretty rare so the reduction to accuracy should be minimal.

Since the notifications pages are hidden for most users does the time the notification was read at currently have any useful data? If not, that might have to wait till we expose the notifications page and collect enough data for it.

**Comment by pushcx** [comment]
@thomasdziedzic Caught a couple small issues:

`_message`: shows `body`, not `markeddown_body`

@veqqq caught that it shows sent messages as unread.

**Comment by thomasdziedzic** [comment]
> `_message`: shows `body`, not `markeddown_body`

Did you mean [linkified_body](https://github.com/lobsters/lobsters/blob/9810c59623dc93a1bf3a2f056705a93cb07e1d68/app/models/message.rb#L105-L107) instead of markeddown_body?

> [@veqqq](https://github.com/veqqq) caught that it shows sent messages as unread.

I couldn't reproduce notifications getting produced for sent messages: Only the [recipient of the message](https://github.com/lobsters/lobsters/blob/9810c59623dc93a1bf3a2f056705a93cb07e1d68/app/jobs/notify_message_job.rb#L11) should be getting notifications. The sender shouldn't be getting a notification.

Can you elaborate this bug above?

**Comment by thomasdziedzic** [comment]
Talked with @pushcx on the stream:
1. I fixed the linkified_body usage. 
2. Since both Pushcx & I don't know the details of the bug that @veqqq mentioned onstream, Pushcx mentioned he would be ok shipping the PR with the bug. @veqqq feel free to add detailed info about the bug you saw.

**Comment by pushcx** [comment]
@thomasdziedzic The `Message` model has its own `has_been_read` tracking that we should replace with the read tracking of the associated `Notification`. Do you mind making this a separate PR from #1649? It feels like a bug to see a message on `/notifications`, click over to it, reply, return to `/notifications`, and still see `(unread)`.

**Comment by mvuets** [comment]
A quick observation about the **/notifications** page. I have an item there that reads "The **author removed** this comment.", however the same item on a story page is rendered as "[Comment **removed by moderator** pushcx: ...]".

**Comment by thomasdziedzic** [comment]
> The `Message` model has its own `has_been_read` tracking that we should replace with the read tracking of the associated `Notification`. Do you mind making this a separate PR from [#1649](https://github.com/lobsters/lobsters/pull/1649)? It feels like a bug to see a message on `/notifications`, click over to it, reply, return to `/notifications`, and still see `(unread)`.

@pushcx Yes, let's link the unread status between notifications for messages and the messages themselves. I'll open a separate PR for it.

The current functionality around message.has_been_read includes marking it as true when showing it in the messages controller, but also keeping it as new by marking it false in the messages controller. The notifications#all action currently doesn't set the read_at only notifications#unread does that. So I'll work through some scenarios:

Scenario 1:
1. User receives message
2. Notification gets created for User
3. User visits the message, this marks message.has_been_read. Proposed: This also marks the notification.read_at = now.
4. The message has_been_read and notification.read_at is in sync

Scenario 2:
1. User receives message
2. Notification gets created for User
3. User visits /notifications. Proposed: This also marks the notification.read_at along with the messages.has_been_read
4. The message has_been_read and notification.read_at is in sync

In Scenario 2, we would have to start marking notification.read_at when visiting /notifications, which is only currently set in /notifications/unread.

Does this sound right? Another thing to keep in mind is notifications get created async from comments/messages, so there is a period of time the the message/comment will live without a notification (could be small, not sure what the mean time to complete Notify{Comment,Message}Job is). 

> A quick observation about the /notifications page. I have an item there that reads "The author removed this comment.", however the same item on a story page is rendered as "[Comment removed by moderator pushcx: ...]".

@mvuets good catch

@pushcx this is due to the logic not being fine grained enough in the [notifications/all page](https://github.com/lobsters/lobsters/blob/bff4466954696fbb21b03dd18d0f86d172e106ae/app/views/notifications/all.html.erb#L34-L53) right now. We did discuss this section as being problematic in the past, should I just remove all that logic and provide some kind of placeholder? Let me know what you think.

**Comment by veqqq** [comment]
> [@veqqq](https://github.com/veqqq) caught that it shows sent messages as unread.

This was a miscommunication, the bug was far worse. But it's better now!

For reference, many messages and comment were unread, including those I replied to, even when replying with the notification's reply interface. That now seems fine.

There is a still a bug where at least one currently open (but not yet responded to) message marked unread. I believe the tab was open for at least 30 minutes. Reloading notifications made it read.

**Comment by veqqq** [comment]
Sorry, scratch that. In "all notifications", the message I mentioned just above(and a few comments) reverted to unread, even though I answered them many hours ago. Reloading "all notifications" does not make them read, only clicking "unread notifications" does.

**Comment by pushcx** [comment]
On placeholders - we've talked about this on the last couple streams and @thomasdziedzic is going to write a PR to remove them. Even if they are correct (and good catch that they were bugged, @mvuets) they are a little puzzle that leads people back into a bad situation, exactly the opposite of what we want them to do.

Flags are pretty rare so I'd rather the page maybe quietly show fewer than 25 notifications instead of defeating the logic around flamewar preventing.

**Comment by thomasdziedzic** [comment]
> A quick observation about the **/notifications** page. I have an item there that reads "The **author removed** this comment.", however the same item on a story page is rendered as "[Comment **removed by moderator** pushcx: ...]".

@mvuets this was fixed in https://github.com/lobsters/lobsters/pull/1684 by just hiding the notifications that were supposed to be not displayed.

**Comment by thomasdziedzic** [comment]
> Sorry, scratch that. In "all notifications", the message I mentioned just above(and a few comments) reverted to unread, even though I answered them many hours ago. Reloading "all notifications" does not make them read, only clicking "unread notifications" does.

@veqqq All notifications now makes the page of displayed unread notifications be marked as read. In addition, it [syncs the read status on the notification with the message now](https://github.com/lobsters/lobsters/pull/1683).

**Comment by mvuets** [comment]
Thanks for the fix, @thomasdziedzic! One more observation about the **/notifications** page: a parent **div** of an **upvoter** link is missing the class **upvoted** when applicable. IOW, upvote arrow links are not highlighted when in fact corresponding comments are upvoted (as it can be seen on other pages).

**Comment by thomasdziedzic** [comment]
> One more observation about the **/notifications** page: a parent **div** of an **upvoter** link is missing the class **upvoted** when applicable. IOW, upvote arrow links are not highlighted when in fact corresponding comments are upvoted (as it can be seen on other pages).

@mvuets I could repro this issue and have opened https://github.com/lobsters/lobsters/pull/1688 to fix it.

**Comment by pushcx** [comment]
OK, to draw a line under it: the big preceeding conversation is about @thomasdziedzic removing a database view as part of #539. He is not continuing to work on this feature request, which is still open and still desired.

We probably want to use the [noticed](https://github.com/excid3/noticed) gem to get a lot of these features.

**Comment by veqqq** [comment]
Issue/feature request: Default functionality changed for posts. It seems one doesn't receive any notifications unless toggling "Submitter is following this story to receive replies via e-mail and Pushover" (which thankfully doesn't send emails, probably due to my settings, but is incorrectly named(?)). However, if toggled, you receive a notification for every comment, including people replying to others.

Previously, one would receive notifications responding to posts/comments, without toggling anything. Losing this is rather annoying. Could you reinstate it somehow? ^^;

Also, it seems like you no longer get notifications of responses to stories on topics you've filtered out.

**Comment by mvuets** [comment]
I have a similar observation and complain. I have posted very few stories in the past and they got even fewer comments, so I can't be certain. However, just recently I didn't tick "Follow" when submitting, because its unfortunate wording suggested that my "Comment Reply: Receive E-mail = off" setting might be disregarded. And later I regretted it: comments came, but both /inbox and /threads pages didn't display them. I usually follow my discussions on /threads, so now I had to remember to check my story explicitly. I just saw this setting "Submitted Story Comment Settings: Show in Your Threads" and it's on, so this adds up to the confusion.

So, bullet point thoughts:
- Rephrase "Submitter is following this story to receive replies via e-mail and Pushover" to "Submitter is following this story to receive notifications" (optionally plus "as per account settings").
- Don't care that much whether "Follow" is on or off by default.
- There is no way to follow/subscribe to my own story after a non-follow submission.
- Consider whether send notifications to all vs. top-level only comments (on submitted stories).
- "Show in Your Threads" setting may be broken.

**Comment by caius** [comment]
Ah I've noticed this too, the badge at the top isn't showing me unread for the week/end threads anymore which I liked. I don't have the "Submitter is following this story" checkbox enabled.

**Comment by dzwdz** [comment]
> It tracks 'unread' separately from the existing replies pages, so clearing one won't clear the other. This is a little confusing while both systems exist, sorry.

To be clear, this will get fixed eventually, right? It's not the intended behavior?

**Comment by veqqq** [comment]
> To be clear, this will get fixed eventually, right? It's not the intended behavior?

I thought it's been "fixed" in that the old system was removed.

**Comment by pushcx** [comment]
Yes, the old system has been removed, so there should be one consistent notion of 'unread'. There isn't a feature to mark a comment or message as 'unread', so if you read it, it should stay read.

**Comment by veqqq** [comment]
Another issue, it gives notification numbers that people have responded to your comments, but it doesn't show the comments themselves if it's on a story whose tag you've filtered.

**Comment by pushcx** [comment]
I've filed #1707 to track this bug with missing onsite notifications when you submit a story.

And then #1708 to track the bug that filtering isn't respected in the count at the top of the page. That's going to be tricky, given that it's happening on the Ruby side instead of in the database. Not considering it is a significant oversight when we were deciding how that should be implemented.

**Comment by pushcx** [comment]
In 917f0a1c I made the reply/mention notifications respect story hiding. A user had a bad time recently because they were repeatedly mentioned in a thread they hid that was about how someone was really shitty to them. It wasn't clear from `/settings` that they could've turned off reply/mention notifications, and it's a failure that the obvious thing they tried didn't work.

We should probably also have a user setting for whether reply/mention respect tag filters.
